staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -126,6 +127,10 @@ func New(
 		lastSyncedNodes: make(map[string][]*v1.Node),
 	}
 
+	if err := s.init(); err != nil {
```

Comment:
yeah, do we have the same bug in the other controllers?

Reply:
other cloud controllers which are used by kcm have checked it in `New` func

Reply:
- https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L81

Reply:
https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L116 (not used by kcm)

Reply:
https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go#L72 doesn't have a cloud field.

Reply:
yes, we do, it turns out some controllers do check if there is a cloud provider not nil on the start functions, see

https://github.com/kubernetes/kubernetes/blob/6cc99274872264a231c61ea373fb0bed1d9e2331/cmd/kube-controller-manager/app/core.go#L129-L133

https://github.com/kubernetes/kubernetes/blob/6cc99274872264a231c61ea373fb0bed1d9e2331/cmd/kube-controller-manager/app/core.go#L297-L300

In addition to this change, let's move those checks to that place too

```
iff --git a/cmd/kube-controller-manager/app/core.go b/cmd/kube-controller-manager/app/core.go
index b3637f273af..164323009c9 100644
--- a/cmd/kube-controller-manager/app/core.go
+++ b/cmd/kube-controller-manager/app/core.go
@@ -92,6 +92,10 @@ func newServiceLBControllerDescriptor() *ControllerDescriptor {
 }
 
 func startServiceLBController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
+       if controllerContext.Cloud == nil {
+               logger.Info("Warning: node-controller is set, but no cloud provider specified. Will not configure node lifecyle controller.")
+               return nil, false, nil
+       }
        serviceController, err := servicecontroller.New(
                controllerContext.Cloud,
                controllerContext.ClientBuilder.ClientOrDie("service-controller"),
@@ -261,6 +265,10 @@ func newCloudNodeLifecycleControllerDescriptor() *ControllerDescriptor {
 
 func startCloudNodeLifecycleController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
        logger := klog.FromContext(ctx)
+       if controllerContext.Cloud == nil {
+               logger.Info("Warning: node-controller is set, but no cloud provider specified. Will not configure node lifecyle controller.")
+               return nil, false, nil
+       }
        cloudNodeLifecycleController, err := cloudnodelifecyclecontroller.NewCloudNodeLifecycleController(
                controllerContext.InformerFactory.Core().V1().Nodes(),
                // cloud node lifecycle controller uses existing cluster role from node-controller
```

Reply:
added.